Feature/128 add html handler

ctmm opened the pull request:
closes #128

timomayer commented:
@ctmm can you please rebase this PR after merging #145 #146

ctmm commented:
@timomayer @SheepFromHeaven, I have updated this branch to reflect the latest changes on master. Ready for review.

tiagomapmarques commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@biotope/resource-loader",
-  "version": "2.1.1",
+  "version": "3.0.0-rc1",
```
This should be done in a separate PR, no?

ctmm replied:
fixed

tiagomapmarques commented on src/ResourceLoader.ts:
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
+		this.mutationObserver.observe(this.options.container, this.mutationObserverConfig);
+	}
 
+	private onMutation(mutationRecordArray: MutationRecord[]): void {
```
NIT: IMO should be `handleMutation`

ctmm replied:
Updated

tiagomapmarques commented on src/ResourceLoader.ts:
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
+	}
 
+	private onMutation(mutationRecordArray: MutationRecord[]): void {
+		mutationRecordArray.forEach(((record: MutationRecord) => {
```
NIT: `record` could be de-constructed

ctmm replied:
Sure. Updated

tiagomapmarques commented on src/ResourceLoader.ts:
```diff
@@ -30,14 +37,33 @@ class ResourceLoader {
+		mutationRecordArray.forEach(((record: MutationRecord) => {
+			record.addedNodes.forEach((node: Node) => {
+				if (node instanceof HTMLElement && node.getAttribute(this.options.resourceListAtrributeSelector)) {
+					this.options = {
```
if `this` refers to the instance of the class, then we should just be adding the `container` attribute, not rewriting the `options` variable altogether (unless there's a valid reason to do it)

ctmm replied:
Good point. Will fix.

tiagomapmarques commented on src/handlers/htmlHandler.ts:
```diff
@@ -0,0 +1,23 @@
+import { HandleOptions } from './../types';
+import { Handler } from '../types';
+
+const isHtml = (options: HandleOptions): boolean => options.resource.path.indexOf('.html') > -1;
```
This will return `true` for this file `file.html.js`, correct? Consider using `slice(-5)` (or `slice(-'.html'.length)`).

ctmm replied:
I see your point, will update.

I followed this convention because it was already in use for similar .js and .css asset handlers.
We should open a new ticket to update the other handlers.

tiagomapmarques commented on src/handlers/htmlHandler.ts:
```diff
@@ -0,0 +1,23 @@
+const isHtml = (options: HandleOptions): boolean => options.resource.path.indexOf('.html') > -1;
+
+const onHtmlLoaded = (options: HandleOptions): void => {
+  const elements: HTMLElement[] = [].slice.call(options.resource.elements);
```
Do we need `[].slice.call`? I think `NodeList` has a `forEach` function…

ctmm replied:
Correct, but the last time I checked Babel did not see fit to poylfill "dom stuff"(their words).
I cast the ```JS NodeList``` into an ```JS Array``` precisely for that reason.

IE11 does not implement the `forEach` method on ```JS NodeList```.

Would prefer to keep as is.

tiagomapmarques commented on src/handlers/htmlHandler.ts:
```diff
@@ -0,0 +1,23 @@
+  const elements: HTMLElement[] = [].slice.call(options.resource.elements);
+
+  options.response.text()
+    .then(stringOfHtml => {
```
NIT: non-essential new line on the `.then`

ctmm replied:
Ok, will change.

tiagomapmarques commented on src/handlers/htmlHandler.ts:
```diff
@@ -0,0 +1,23 @@
+const onHtmlLoaded = (options: HandleOptions): void => {
+  const elements: HTMLElement[] = [].slice.call(options.resource.elements);
+
+  options.response.text()
```
Can we return this promise?
(If we need it to be a somewhat "synced" function, we can't do it if we don't return this line)

ctmm replied:
I am not seeing a quick way to introduce said functionality because the current logic expects handlers to never return anything.

I haven't yet invested a lot of time on a possible implementation, but it strikes me as something that would take a minimum of half a day effort.

Would this be a deal breaker for merging this PR or is it something we can create an issue for and address later?

tiagomapmarques replied:
Addressing it later is fine :)

tiagomapmarques commented on src/defaultOptions.ts:
```diff
@@ -1,6 +1,7 @@
 import cssHandler from './handlers/cssHandler';
 import jsHandler from './handlers/jsHandler';
 import { ResourceLoaderOptions } from './types';
+import htmlHandler from './handlers/htmlHandler';
```
NIT: should be together with L1 and L2

ctmm replied:
Fixed

ctmm commented:
@tiagomapmarques I addressed all of your comments, updating the code for small changes. For these small changes I took the liberty of marking the conversation as resolved.

All remarks you made that are open for discussion I left open. 
Kindly ask you to get back at me for the open topics.

Thank you!
